### PR TITLE
add enableNodePublicIP to managed machine pools

### PR DESCRIPTION
**Pull request by nojnhuh**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**: This PR adds the ability for users to define AKS's `enableNodePublicIP` on AzureManagedMachinePools to give nodes externally available IP addresses.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Part of #2266

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
Implement EnableNodePublicIP for AzureManagedMachinePool
```

### CODE REVIEW

File: exp/api/v1beta1/azuremanagedmachinepool_webhook.go [hunk 1]
```diff
@@ -336,3 +323,21 @@ func ensureStringSlicesAreEqual(a []string, b []string) bool {
 	}
 	return true
 }
+
+func validateBoolPtrImmutable(path *field.Path, oldVal, newVal *bool) *field.Error {
```
Comment (jackfrancis) [inline exp/api/v1beta1/azuremanagedmachinepool_webhook.go:327]: Nice

File: exp/api/v1beta1/azuremanagedmachinepool_webhook.go [hunk 2]
```diff
@@ -197,32 +197,19 @@ func (m *AzureManagedMachinePool) ValidateUpdate(oldRaw runtime.Object, client c
...
-					m.Spec.EnableUltraSSD,
-					"field is immutable, unsetting is not allowed"))
-		}
+	if err := validateBoolPtrImmutable(
+		field.NewPath("Spec", "EnableUltraSSD"),
+		old.Spec.EnableUltraSSD,
+		m.Spec.EnableUltraSSD); err != nil {
+		allErrs = append(allErrs, err)
+	}
+	if err := validateBoolPtrImmutable(
```
Comment (jackfrancis) [inline exp/api/v1beta1/azuremanagedmachinepool_webhook.go:206]: I don't actually see specific documentation that says that this node pool configuration is create-time only.

https://docs.microsoft.com/en-us/azure/aks/use-multiple-node-pools#assign-a-public-ip-per-node-for-your-node-pools

Did we confirm that you definitely can't change this configuration after the node pool is already created?

Comment (nojnhuh) [reply exp/api/v1beta1/azuremanagedmachinepool_webhook.go:206]: Yeah, when I try to edit that field I see errors in the logs like
```
E0908 20:44:39.439710     103 controller.go:326]  "msg"="Reconciler error" "error"="error creating AzureManagedMachinePool default/pool1: failed to reconcile machine pool pool1: failed to create resource aks-jon/pool1 (service: agentpools): failed to begin operation: Code=\"PropertyChangeNotAllowed\" Message=\"Changing property 'agentPoolProfile.enableNodePublicIP' is not allowed.\" Target=\"agentPoolProfile.enableNodePublicIP\"" "azureManagedMachinePool"={"name":"pool1","namespace":"default"} "controller"="azuremanagedmachinepool" "controllerGroup"="infrastructure.cluster.x-k8s.io" "controllerKind"="AzureManagedMachinePool" "name"="pool1" "namespace"="default" "reconcileID"="5c493f37-3585-40b3-ac0a-b93b548ab047" 
```

Comment (nojnhuh) [reply exp/api/v1beta1/azuremanagedmachinepool_webhook.go:206]: That same error shows up in both cases when it was created as true or false and toggling to the other.

Comment (jackfrancis) [reply exp/api/v1beta1/azuremanagedmachinepool_webhook.go:206]: Great, that settles it!

**Review by jackfrancis — APPROVED** [review]
/lgtm
/approve

**Pull request merged**